### PR TITLE
[TRTLLM-15406][feat] Support occurrence penalties in one-model speculative decoding

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Dev Engineer Review

- Added occurrence-penalty support for one-model speculative decoding modes.
- Added deploy-time `enable_penalty` configuration.
- Rejects penalty requests when support is disabled.
- Rejects penalties for tree-based speculative decoding.
- Added CUDA-graph-safe penalty state, prompt seeding, row mapping, and accepted-token updates.
- Applies penalties only to cloned target logits. Draft distributions remain unchanged.
- Centralized penalty arithmetic in `occurrence_penalized_logits`.
- Added documentation and golden-manifest coverage.
- No test files or test-list files were modified.

## QA Engineer Review

No test changes.

## Description

One-model speculative decoding (MTP, Eagle3 one-model, DSpark, DFlash, DraftTarget, PARD) silently ignores `repetition_penalty`, `presence_penalty` and `frequency_penalty`. The worker's fused sampling path only consumes temperature/top_k/top_p (`SpecMetadata.populate_sampling_params_for_one_model`), so a request asking for any of the three decodes from an unpenalized distribution with no diagnostic. `min_p` hits the same wall and is already rejected explicitly in `SpecSampler.validate_request`; the penalties had no such guard.

This PR adds a deploy-time `enable_penalty` flag on `DecodingBaseConfig`:

- **When set**, the occurrence workspace is preallocated on `SpecMetadata` (CUDA graphs capture fixed buffer addresses, so it cannot be allocated lazily on first use the way `TorchSampler.PenaltyStore.ensure_workspace` does), and the penalties are applied to the target logits in `SpecWorkerBase._accept_draft_tokens` — ahead of both the strict and the rejection acceptance branch, so both verify against the penalized distribution.
- **When unset**, a request carrying any of the three penalties is rejected at admission rather than served incorrectly, following the `min_p` precedent (raising from `validate_request` fails only the offending request instead of the whole executor step).

Four design points worth calling out:

1. **Occurrence state is split at `prompt_ignore_length`.** The ignored prompt prefix lives in a packed bitmask (1 bit per id) and drives repetition only; everything else — the rest of the prompt plus every generated token — lives in an int32 count tensor and drives all three penalties. This mirrors the C++ `batchApplyPenalty` kernel and the two-model `TorchSampler` path, so the same request decodes the same way whichever backend serves it. The bitmask is viable precisely because repetition only asks "seen or not", which is all the ignored prefix contributes.

2. **PARD shares the linear row mapping.** The apply pass is hooked *after* `_reshape_logits_for_accept`, which already narrows PARD's `2K` rows per gen request to `K+1`, so no per-mode special case is needed.

3. **Both penalty implementations live in `sampler/penalties.py`, sharing one
formula.** The eager `PenaltyHandler` and the CUDA-graph-safe `PenaltyState` /
`apply_penalties` pair sit side by side there rather than in separate packages:
nothing about the latter is specific to speculation, and a future graph-capturing
`TorchSampler` path can reuse it as is.

   The arithmetic is not duplicated between them — both call
`ops.vanilla.occurrence_penalized_logits`, which takes the operands already gathered
and knows nothing about slots, beams or row layouts. (It carries separate masks for
repetition and for presence/frequency, because a prompt prefix excluded by
`prompt_ignore_length` is "seen" for the former but not the latter.) Previously each
side held its own copy of the same eight lines, kept aligned only by a
keep-in-sync comment.

   What is *not* merged is the state management, because the execution contracts
differ: the eager half may allocate lazily and skip work with a host-side branch,
while the graph-safe half must preallocate at stable addresses and gate on device
tensors, and only the eager half carries the per-beam bookkeeping (speculative
decoding is always beam width 1). The module docstring records both halves and why
each exists.

4. **Tree modes are not supported yet** and are rejected at admission. A tree's rows are nodes whose history is the root path rather than the rows before them, so the linear lower-triangular prefix would let sibling branches penalize each other. Requests that do *not* use penalties are unaffected on tree modes.

The draft distribution is deliberately left unpenalized (matching vLLM). Rejection sampling only requires `draft_probs` to match how the draft tokens were actually drawn, so it stays unbiased; the cost is acceptance rate, not correctness.

Within a step, position `k` is penalized against positions `0..k-1` of that same step — those tokens are not in the persistent counts yet, since only accepted tokens are committed at step end.

## Defects fixed during bring-up

Bugs found while validating the feature end-to-end and during review, folded
into this PR:

1. **Out-of-bounds scatter (crash).** The intra-step prefix scattered draft token
   ids into a `[rows, vocab_size]` workspace after masking only for `id >= 0`. A
   draft slot can legitimately hold an id at or above the target vocab -- an
   unfilled/padding entry, or a reduced-vocab draft head -- which tripped
   `scatter gather kernel index out of bounds`. Ids are now bounded on both sides
   and out-of-range entries are dropped from the valid mask.

2. **Per-sequence state reset on every prefill chunk.** `is_context_init_state`
   stays true for every chunk of a chunked prefill, so gating the reset on it
   alone re-cleared the occurrence counts and re-seeded the prompt bitmask on each
   chunk, discarding what earlier chunks had built. Now also gated on
   `is_last_context_chunk`, matching how `guided_decoder` pairs the two.

3. **In-place penalty leaked into the reported logits.** The worker keeps a
   reference to the logits it passes to acceptance and returns that same tensor as
   the step's `logits` output (Eagle3 aliases it as `raw_logits`), so penalizing in
   place fed penalized scores to logprobs and other consumers that should see the
   model's own values. The penalties are now applied to a clone.

4. **`prompt_ignore_length` was ignored.** The whole prompt was recorded in the
   bitmask, which silently implemented `prompt_ignore_length = prompt length`
   regardless of what the request asked for, and disagreed with both the C++ kernel
   and the two-model path. The prompt is now split at N as documented.

5. **Intra-step mapping was quadratic in batch size.** The prefix tensors were
   `[total_rows, total_rows]` with a block-diagonal write pattern, but a row only
   ever reads its own request's `draft_len` positions — roughly 512 MiB + 64 MiB per
   step at 2048 generation requests with `draft_len=3`. They are now
   `[total_rows, draft_len]` (48 KiB instead of 32 MiB at 512 requests), and
   cross-request isolation holds by construction rather than by masking.

6. **Non-finite logits were clamped away.** The formula clamped every logit to the
   finite range, including ones that arrived non-finite. Embedding bias runs before
   penalties, so a token the user had masked out with `-inf` came back as the finite
   minimum and became eligible for sampling again. Only values the penalty arithmetic
   itself could push out of range are clamped now; non-finite inputs pass through
   unchanged. This affected the eager path as well, since both share the formula.

7. **Out-of-range slot indices were not rejected.** The prompt-seeding guards checked
   only the upper bound, so a negative slot would wrap onto another request's row and
   seed its occurrence history. Both the op and its caller now accept live rows only.

8. **The padding scratch row was not published without rejection sampling.**
   `dummy_slot_row` — where CUDA-graph padding requests are routed — is set by
   `prepare_rejection_sampling_buffers`, which returns early when rejection sampling
   is off. With `enable_penalty=True` and `use_rejection_sampling=False` it kept its
   default of 0, a live request's row, so padding requests would have polluted a real
   sequence's occurrence history. It is now published from `prepare_penalty_buffers`
   too.

## Test Coverage

Unit tests (added, run on H200):

- penalty formulas vs. the `TorchSampler` reference (`Fusions._apply_occurrence_penalties_impl`), intra-step prefix, inactive/disabled no-ops
- prompt split at `prompt_ignore_length`: the ignored prefix drives repetition only, the counted remainder drives all three; cross-word bit packing, out-of-range id filtering (multimodal placeholders), idempotence on repeated ids
- row mapping: `[ctx 1, gen K+1]` layout, cross-request isolation, PARD `2K` narrowing, context-only batches, and a regression check that the intra-step tensors stay `[rows, draft_len]` rather than growing with the batch
- per-slot parameter population: slot reuse clears stale state, context requests reset counts, dummy requests skipped
- admission: flag on/off × all three penalties, tree-mode rejection, `min_p` still rejected

The penalty tests use distinct, non-default initial logits and assert that columns
which should not change are bit-identical afterwards, so an accidental write cannot
hide behind a zero initial value.

End-to-end (Eagle3 one-model, Llama-3.1-8B + EAGLE3-LLaMA3.1-Instruct-8B, H200):

- `enable_penalty=False` + `repetition_penalty` → request rejected with an actionable message; unpenalized requests unaffected
- `enable_penalty=True` → penalties measurably change output (distinct-word ratio 0.025 → 1.000 on a repetition-inducing prompt); `frequency_penalty` likewise
- **CUDA graphs on vs. off → byte-identical output**
- in a mixed batch, the unpenalized member is byte-identical to its solo run

Regression: `tests/unittest/_torch/sampler` passes 257 tests, which covers the eager
`TorchSampler` penalty path that now shares the extracted formula;
`tests/unittest/_torch/speculative/` shows no change vs. the base commit (the failures
present are pre-existing and require model downloads unavailable in this environment);
`tests/unittest/api_stability` unchanged.

## Notes on batched determinism

An earlier revision of this description listed a mixed-batch difference as an open
bug: a penalized request produced different text when batched with an unpenalized
one. That has since been investigated and is **not introduced by this PR** — the
same divergence reproduces with speculative decoding switched off entirely, on the
existing `TorchSampler` penalty path. It is a pre-existing property of batched
penalized decoding: the repetition penalty divides logits, batching perturbs
reduction order slightly, and a greedy chain amplifies a near-tie into a different
continuation. Unpenalized requests sharing the batch remain byte-identical, and
accepted-token counts match.

Verified step by step with CUDA graphs disabled: occurrence counts, intra-step
prefixes, row->slot mapping, accepted-token counts and accepted token ids are all
identical between the alone and mixed cases, and the apply op is bit-exact across
batch sizes.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/interface.py [hunk 1]
```diff
@@ -1526,6 +1758,46 @@ def _sample_and_accept_draft_tokens_base(
 
         return accepted_tokens, num_accepted_tokens
 
+    def _apply_occurrence_penalties(self, logits, draft_tokens, num_contexts,
+                                    batch_size, spec_metadata):
```
Comment (lori-ren) [inline tensorrt_llm/_torch/speculative/interface.py]: Miss type hints for params, which will fail mypy checks.

File: tensorrt_llm/_torch/speculative/penalty_ops.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        # Each gen row writes into its own request's column block, so requests can
+        # never read each other's tokens.
+        col_base = num_contexts + g_of_row * rows_per_gen
+        cols = col_base.unsqueeze(1) + torch.arange(draft_len, device=device)
+        rows = (num_contexts + gen_rows).unsqueeze(1).expand(-1, draft_len)
+        # Column j counts for row k iff j < k: strictly earlier positions only.
+        valid = torch.arange(draft_len, device=device).unsqueeze(0) < k_of_row.unsqueeze(1)
+        vals = drafts[g_of_row]
+        intra_tokens[rows.reshape(-1), cols.reshape(-1)] = vals.reshape(-1)
+        intra_valid[rows.reshape(-1), cols.reshape(-1)] = valid.reshape(-1)
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/penalty_ops.py]: Reduced to `[total_rows, draft_len]` in d4b0519 - 48 KiB instead of 32 MiB at 512 requests.

File: tensorrt_llm/_torch/speculative/interface.py (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/interface.py]: Added in d4b0519, thanks.

**Comment by zhaoyangwang-nvidia** [comment]
@lori-ren you own the `TorchSampler` penalty code, so flagging this: I've unified what's shareable between it and the CUDA-graph-safe implementation this PR adds.

- The formula is now one function, `ops.vanilla.occurrence_penalized_logits`, called by both (338b0315) — previously the same eight lines twice. It takes two separate masks, since a prefix excluded by `prompt_ignore_length` is "seen" for repetition but not presence/frequency.
- Both implementations now sit in `sampler/penalties.py` (e93599b3), so a future graph-capturing `TorchSampler` path can reuse the new half.

State management is deliberately not merged — yours allocates lazily and carries the per-beam bookkeeping; the graph-safe half must preallocate and gate on device tensors. `tests/unittest/_torch/sampler` passes 257 tests, but you'd catch a semantic drift faster than they would. Mind taking a look?

**Comment by lori-ren** [comment]
> @lori-ren you own the `TorchSampler` penalty code, so flagging this: I've unified what's shareable between it and the CUDA-graph-safe implementation this PR adds.
> 
> * The formula is now one function, `ops.vanilla.occurrence_penalized_logits`, called by both ([338b031](https://github.com/NVIDIA/TensorRT-LLM/commit/338b03154b10ab8b017a69b6aaade5ed19d6939c)) — previously the same eight lines twice. It takes two separate masks, since a prefix excluded by `prompt_ignore_length` is "seen" for repetition but not presence/frequency.
> * Both implementations now sit in `sampler/penalties.py` ([e93599b](https://github.com/NVIDIA/TensorRT-LLM/commit/e93599b3f0228f3882a7720ff4e69379ea18a32c)), so a future graph-capturing `TorchSampler` path can reuse the new half.
> 
> State management is deliberately not merged — yours allocates lazily and carries the per-beam bookkeeping; the graph-safe half must preallocate and gate on device tensors. `tests/unittest/_torch/sampler` passes 257 tests, but you'd catch a semantic drift faster than they would. Mind taking a look?

LGTM. Only one non-blocking point: `ops.vanilla.occurrence_penalized_logits` is intended to be a private op which is owned and managed by pysampler. However, your changes to one-model spec decode sampler refer to the private op, which makes the ownership unclear. It's okay for the moment, but better we should reconsider where to place these ops.

File: tensorrt_llm/_torch/pyexecutor/sampler/penalties.py [hunk 3]
```diff
@@ -618,3 +637,430 @@ def _arm_pending_tokens(
...
+    disturb a live request's history.
+    """
+
+    counts: torch.Tensor
+    prompt_mask: torch.Tensor
+    repetition: torch.Tensor
+    presence: torch.Tensor
+    frequency: torch.Tensor
+    active: torch.Tensor
+    dummy_row: int
```
Comment (lori-ren) [inline tensorrt_llm/_torch/pyexecutor/sampler/penalties.py:709]: I couldn't find any reference to this field. What is it for?

File: tensorrt_llm/_torch/pyexecutor/sampler/penalties.py [hunk 4]
```diff
@@ -618,3 +637,431 @@ def _arm_pending_tokens(
...
+    No-op unless the penalty buffers exist (i.e. ``enable_penalty``).
+    """
+    if not getattr(spec_metadata, "enable_penalty", False):
+        return
+    state = spec_metadata.penalty_state
+    if state is None or prompt_tokens.numel() == 0:
+        return
+    prompt_mask, counts = state.prompt_mask, state.counts
+    if slot >= prompt_mask.size(0):
+        return
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/penalties.py]: Both the op and its caller now accept live rows only, in 5dda6ec.

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py [hunk 5]
```diff
@@ -187,6 +187,49 @@ def sample_rejected(
...
+    penalized = logits.float()
+    repeated = torch.where(penalized < 0, penalized * repetition, penalized / repetition)
+    penalized = torch.where(seen_for_repetition, repeated, penalized)
+    penalized = penalized - torch.where(
+        seen_for_presence,
+        presence + frequency * count.to(torch.float32),
+        penalized.new_zeros(()),
+    )
+    limit = torch.finfo(logits.dtype).max
+    return penalized.clamp(-limit, limit).to(logits.dtype)
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py]: Fixed in 5dda6ec - non-finite inputs now pass through unclamped, with a regression test.

**Comment by zhaoyangwang-nvidia** [comment]
Thanks! Worth noting `sampler/ops` already serves speculative decoding today — `flashinfer.py` has a "Speculative-decoding samplers" section with `sampling_batch_spec_dec_one_model_for_rejection`, `vanilla.py` has `get_rejected_indices`, and `_torch/speculative/` already imports `greedy_search_sampling_batch` and `GREEDY_TEMPERATURE_THRESHOLD` from there. Also, after e93599b the graph-safe half lives in `sampler/penalties.py`, so the `occurrence_penalized_logits` caller is inside pysampler rather than in the spec package.

File: tensorrt_llm/_torch/pyexecutor/sampler/penalties.py (hunk 3 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/penalties.py:709]: Good catch — it was unreferenced because the routing read `SpecMetadata.dummy_slot_row` instead, which is only published by `prepare_rejection_sampling_buffers`. With penalties on and rejection sampling off it stayed 0, so CUDA-graph padding requests were routed onto a live request's row. Fixed in ed7d592 with a regression test.

**Review by lori-ren — APPROVED** [review]

**Review by juney-nvidia — APPROVED** [review]
Approved from API/doc/OSS compliance perspectives.
